File: src/main/java/algo/depthFirstSearch/KeysAndRooms_841.java
package algo.depthFirstSearch;

import java.util.ArrayList;
import java.util.List;

public class KeysAndRooms_841 {
    public static boolean canVisitAllRooms(List<List<Integer>> rooms) {
        var len = rooms.size();
        var visited = new boolean[len];

        visitRooms(rooms, 0, visited);

        for (var v : visited) {
            if (!v) {
                return false;
            }
        }
        return true;
    }

    private static void visitRooms(List<List<Integer>> rooms, int id, boolean[] visited) {
        if (visited[id]) {
            return;
        }
        visited[id] = true;
        for (var key : rooms.get(id)) {
            visitRooms(rooms, key, visited);
        }
    }

    public static void main(String[] args) {
        var rooms = new ArrayList<List<Integer>>();
        var room1 = new ArrayList<Integer>();
        room1.add(1);
        var room2 = new ArrayList<Integer>();
        room2.add(2);
        var room3 = new ArrayList<Integer>();
        room3.add(3);
        rooms.add(room1);
        rooms.add(room2);
        rooms.add(room3);
        rooms.add(new ArrayList<>());
        System.out.println(canVisitAllRooms(rooms));
    }
}

File: src/main/kotlin/algo/depthFirstSearch/KeysAndRooms_841.kt
package algo.depthFirstSearch

class KtKeysAndRooms_841 {
    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        fun visit(rooms: List<List<Int>>, room: Int, visited: BooleanArray): Int {
            var num = 1
            if (visited[room]) {
                return 0
            }
            visited[room] = true
            rooms[room].forEach {
                num += visit(rooms, it, visited)
            }
            return num
        }
        return visit(rooms, 0, BooleanArray(rooms.size)) == rooms.size
    }
}

fun main() {
    val visit = KtKeysAndRooms_841()
    println(visit.canVisitAllRooms(listOf(listOf(1), listOf(2), listOf(3), listOf())))
    println(visit.canVisitAllRooms(listOf(listOf(1,3), listOf(3,0,1), listOf(2), listOf(0))))
}